File: 2958. Length of Longest Subarray With at Most K Frequency/maxSubarrayLength.py
from typing import List

class Solution:
  def maxSubarrayLength(self, nums: List[int], k: int) -> int:
    frequencyMap = {}
    res = 0
    
    left = 0
    for right in range(len(nums)):
      if nums[right] not in frequencyMap:
        frequencyMap[nums[right]] = 0
      frequencyMap[nums[right]] += 1
      
      while frequencyMap[nums[right]] > k:
        frequencyMap[nums[left]] -= 1
        left += 1
      
      res = max(res, right - left + 1)
      
    return res    
  
nums = [1, 2, 3, 1, 2, 3, 1, 2]
k = 2
solution = Solution()
print('maxSubarrayLength: ', solution.maxSubarrayLength(nums, k))

File: 2958. Length of Longest Subarray With at Most K Frequency/maxSubarrayLength.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const maxSubarrayLength = (nums, k) => {
  const frequencyMap = {};
  let res = 0;
  let left = 0;

  for (let right = 0; right < nums.length; right++) {
    if (!frequencyMap[nums[right]]) {
      frequencyMap[nums[right]] = 0;
    }
    frequencyMap[nums[right]]++;

    while (frequencyMap[nums[right]] > k) {
      frequencyMap[nums[left]]--;
      left++;
    }

    res = Math.max(res, right - left + 1)
  }

  return res;
};

const nums = [1, 2, 3, 1, 2, 3, 1, 2];
const k = 2;
console.log('maxSubarrayLength: ', maxSubarrayLength(nums, k))
